proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java

```diff
@@ -107,8 +107,11 @@ public void handle(PacketWrapper packet) throws Exception
     @Override
     public void handle(KeepAlive alive) throws Exception
     {
-        server.setSentPingId( alive.getRandomId() );
-        con.setSentPingTime( System.currentTimeMillis() );
+        if ( server.getSentKeepAlives().getFirst().getMillis() + bungee.getConfig().getTimeout() < System.currentTimeMillis() )
```

Comment:
Uh I'm not sure about this part, because the bungee timeout may not match the rate at which the server is sending keepalives

Reply:
This check was not added to rigorously check the bungee timeout, it prevents overflowing the sentKeepAlives collection by forcing the client to answer all keepalives at least in bungee's timeout setting. This can be lazily checked whenever the server sends a new keepalive because otherwise the buffer would not grow.

I don't think we have to do special accounting anywhere for servers which are sending a bunch of keepalives in bungee-timeout-time and then no keepalive for the rest of the connection and the client being malicious and not answering the keepalives.
No server would currently behave like that, there is no valid reason for servers to behave like that. A malicious client would at max cause bungee to save the last bungee-timeout-time of keepalives sent by the server.